src/workloads/query/TimeseriesTSBSQuery.yml

```diff
@@ -0,0 +1,749 @@
+                  [
+                    {
+                      $match:
+                        {
+                          time:
+                            {
+                              $gte: { ^Cycle: { ofLength: 100, fromGenerator: { ^IncDate: { start: *StartDate, step: 2520000 } } } },
+                              $lt: { ^Cycle: { ofLength: 100, fromGenerator: { ^IncDate: { start: *startDatePlusTwelveHours, step: 2520000 } } } }
+                            },
+                          usage_user: { $gt: 90 }, # 90 is hard-coded in TSBS.
+                        },
+                    },
+                    { $set: { tags: "$tags.hostname" } },
+                  ]
+
+# lastpoint computes the last row per host.
+  - Name: lastpoint
+    Type: CrudActor
+    Threads: 8
+    Phases:
+      OnlyActiveInPhases:
+        Active: [15]
+        NopInPhasesUpTo: *maxPhases
+        PhaseConfig:
+          Repeat: *repeat
+          Database: *database
+          Collection: *collection
+          Operations:
+            - OperationName: aggregate
+              OperationCommand:
+                Pipeline:
+                  [
+                    { $sort: { "tags.hostname": 1, "time": -1 } },
+                    {
+                      $group:
+                        {
+                          _id: { "hostname": "$tags.hostname" },
+                          time: { $first: "$time" },
+                          measurement: { $first: "$measurement" },
+                          usage_softirq: { $first: "$usage_softirq" },
+                          musage_steal: { $first: "$usage_steal" },
+                          usage_guest: { $first: "$usage_guest" },
+                          usage_user: { $first: "$usage_user" },
+                          usage_system: { $first: "$usage_system" },
+                          usage_iowait: { $first: "$usage_iowait" },
+                          usage_irq: { $first: "$usage_irq" },
+                          usage_nice: { $first: "$usage_nice" },
+                          usage_idle: { $first: "$usage_idle" },
+                          usage_guest_nice: { $first: "$usage_guest_nice" },
+                        },
+                    },
+                  ]
+
+  # The queries below are not from tsbs_query, but test bucket-targeting. We want to target ~1% of buckets.
+  - Name: match_on_measurements_selective_query
+    Type: CrudActor
+    Threads: 8
+    Phases:
+      OnlyActiveInPhases:
+        Active: [16]
+        NopInPhasesUpTo: *maxPhases
+        PhaseConfig:
+          Repeat: 100
+          Database: *database
+          Collection: *collection
+          Operations:
+            - OperationName: aggregate
+              OperationCommand:
+                Pipeline:
+                  # All usage_* fields are random integers between 0-100.
+                  [
+                    {
+                      $match:
+                        {
+                          $or:
```

Comment:
In my local experiment with the tsbs dataset, this query generates the following filter on the bucket collection:
```
"$or" : [
	{"$and" : [
		{"control.max.usage_nice" : {"$_internalExprGt" : 0}},
		{"control.min.usage_nice" : {"$_internalExprLt" : 5}}
	]},
	{"$and" : [
		{"control.max.usage_guest" : {"$_internalExprGte" : 80}},
		{"control.min.usage_guest" : {"$_internalExprLt" : 85}}
	]}
]
```
And it selects 56,288 buckets out of 161,448, which is **~35%** and not ~1% as you claim in the comment.

I haven't tried all of the measurements but it seems that a single measurement value hits ~20% of the buckets. Assuming independent distributions, I'd expect that $and of three $eq on different measurements would yield ~0.8% and for two measurements would get ~4%. Either is OK for this test (but I'd check against the real data before commiting to a choice of filter).

Reply:
Intresting, so the values for the different "usage_*" metrics aren't uniformly distributed. Well, then yes, we have to double-check the actual selectivities of all queries we use here and be careful when/if changing them. Thank you for running the experiments!